File: projects/tox4j/src/test/java/im/tox/gui/domain/FileTransferModel.java
package im.tox.gui.domain;

import javax.swing.*;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.NoSuchElementException;

public final class FileTransferModel extends AbstractListModel<FileTransfer> {

  private final List<List<FileTransfer>> transfers = new ArrayList<>();

  @Override
  public int getSize() {
    int size = 0;
    for (List<FileTransfer> list : transfers) {
      size += list.size();
    }
    return size;
  }

  @Override
  public FileTransfer getElementAt(int index) {
    int position = 0;
    for (List<FileTransfer> list : transfers) {
      if (position + list.size() > index) {
        return list.get(index - position);
      }
      position += list.size();
    }
    throw new NoSuchElementException(String.valueOf(index));
  }

  private List<FileTransfer> ensureFileNumber(int friendNumber, int fileNumber) {
    while (transfers.size() <= friendNumber) {
      transfers.add(new ArrayList<FileTransfer>());
    }
    List<FileTransfer> list = transfers.get(friendNumber);
    while (list.size() <= fileNumber) {
      list.add(null);
    }
    return list;
  }

  public void addOutgoing(int friendNumber, File file, int fileNumber) {
    List<FileTransfer> list = ensureFileNumber(friendNumber, fileNumber);
    list.set(fileNumber, new FileTransferOutgoing(file));
  }

  public void addIncoming(int friendNumber, int fileNumber, int kind, long fileSize, File file) {
    List<FileTransfer> list = ensureFileNumber(friendNumber, fileNumber);
    list.set(fileNumber, new FileTransferIncoming(file));
  }

  public FileTransfer get(int friendNumber, int fileNumber) {
    return transfers.get(friendNumber).get(fileNumber);
  }

  public void remove(int friendNumber, int fileNumber) throws IOException {
    get(friendNumber, fileNumber).close();
  }

}

File: projects/tox4j/src/test/java/im/tox/gui/domain/FileTransferModel.scala
package im.tox.gui.domain

import java.io.{ File, IOException }
import java.util.NoSuchElementException
import javax.swing.AbstractListModel

import scala.collection.mutable.ArrayBuffer

final class FileTransferModel extends AbstractListModel[FileTransfer] {
  private val transfers = new ArrayBuffer[ArrayBuffer[FileTransfer]]

  override def getSize: Int = {
    var size = 0
    for (list <- transfers) {
      size += list.length
    }
    size
  }

  @SuppressWarnings(Array("org.brianmckenna.wartremover.warts.Return"))
  override def getElementAt(index: Int): FileTransfer = {
    var position = 0
    for (list <- transfers) {
      if (position + list.length > index) {
        return list(index - position) // scalastyle:ignore return
      }
      position += list.length
    }
    throw new NoSuchElementException(String.valueOf(index))
  }

  private def ensureFileNumber(friendNumber: Int, fileNumber: Int): ArrayBuffer[FileTransfer] = {
    while (transfers.length <= friendNumber) {
      transfers += new ArrayBuffer[FileTransfer]
    }
    val list = transfers(friendNumber)
    while (list.size <= fileNumber) {
      list += null
    }
    list
  }

  def addOutgoing(friendNumber: Int, file: File, fileNumber: Int): Unit = {
    val list = ensureFileNumber(friendNumber, fileNumber)
    list(fileNumber) = new FileTransferOutgoing(file)
  }

  def addIncoming(friendNumber: Int, fileNumber: Int, kind: Int, fileSize: Long, file: File): Unit = {
    val list = ensureFileNumber(friendNumber, fileNumber)
    list(fileNumber) = new FileTransferIncoming(file, kind, fileSize)
  }

  def get(friendNumber: Int, fileNumber: Int): FileTransfer = {
    transfers(friendNumber)(fileNumber)
  }

  @throws[IOException]
  def remove(friendNumber: Int, fileNumber: Int): Unit = {
    get(friendNumber, fileNumber).close()
  }
}
